### PR TITLE
Improve pump functionality

### PR DESCRIPTION
**Pull request by akx**
This branch had been lying around for over a year, apparently...

### CODE REVIEW

File: hai/pipe_pump.py [hunk 1]
```diff
@@ -27,20 +27,31 @@ def register(self, key: str, fileobj: Optional[IO[bytes]]) -> None:
...
-                data = fileobj.read(self.read_size)
-                self.feed(key.data, data)
+                return 0
+            for reads in range(repeat):
+                should_repeat = False
+                for (key, _event) in self.selector.select(timeout=timeout):
+                    fileobj = key.fileobj  # type: IO[bytes]  # type: ignore[assignment]
+                    data = fileobj.read(self.read_size)
+                    self.feed(key.data, data)
+                    if repeat:
```
Comment (ruksi) [inline hai/pipe_pump.py:50]: shouldn't this be `if reads` as `repeat` is the function parameter which should always be truelike

**Review by ruksi — CHANGES_REQUESTED** [review]

File: hai/pipe_pump.py (hunk 1 above)
Comment (ruksi) [reply hai/pipe_pump.py:50]: or, no, `range(1)` would give it 0 hmm

Comment (ruksi) [reply hai/pipe_pump.py:50]: a bit confusing logic :thinking:

Comment (akx) [reply hai/pipe_pump.py:50]: I agree. I might revisit it...

File: hai/pipe_pump.py [hunk 2]
```diff
@@ -27,20 +27,32 @@ def register(self, key: str, fileobj: Optional[IO[bytes]]) -> None:
...
+                read_num += 1
+                should_repeat = False
+                for (key, _event) in self.selector.select(timeout=timeout):
+                    fileobj = key.fileobj  # type: IO[bytes]  # type: ignore[assignment]
+                    data = fileobj.read(self.read_size)
+                    self.feed(key.data, data)
+                    should_repeat = True  # Got data, should try again
+                if not should_repeat:
+                    break
+        return read_num
```
Comment (ruksi) [inline hai/pipe_pump.py:55]: excelente!

**Review by ruksi — APPROVED** [review]
lgtm!

**Pull request merged**